sunpy/data/data_manager/cache.py

```diff
@@ -40,56 +39,60 @@ def download(self, urls, namespace='', redownload=False):
-        redownload : `bool`
+            The list is for urls of the same file but from different sources.
+        namespace : `str`, optional
+            A namespace to be used for the file name.
+            Defaults to an empty string.
+        redownload : `bool`, optional
             Whether to skip cache and redownload.
+            Defaults to `False`.
 
         Returns
         -------
-        `pathlib.PosixPath`
+        `pathlib.Path`
             Path to the downloaded file.
         """
-        if isinstance(urls, str):
+        if isinstance(urls, str | Path):
             urls = [urls]
-        # Program flow
-        # 1. If redownload: Download, update cache and return file path
-        # 2. If not redownload: Check cache,
-        #    i. If present in cache:
-        #        - If cache expired, remove entry from cache, download and add to cache
-        #        - If cache not expired, return path
-        details = None
+        # Logic plan
+        # 1. Check if the file is present in cache by url
+        # 2. If present and it has not expired nor redownload, return the file path
+        # 3. If not present or present and (expired or redownload), download the file and update cache
+        #   a. If there is an error from the above steps, we will return the file from the cache if present
         for url in urls:
-            details = self._get_by_url(url)
-            if details:
+            cache_details = self._get_by_url(url)
+            if cache_details:
                 break
-        if details:
-            if redownload or self._has_expired(details):
-                # if file is in cache and it has to be redownloaded or the cache has expired
-                # then remove the file and delete the details from the storage
-                os.remove(details['file_path'])
-                self._storage.delete_by_key('url', details['url'])
-            else:
-                return Path(details['file_path'])
-
-        file_path, file_hash, url = self._download_and_hash(urls, namespace)
-
-        self._storage.store({
-            'file_hash': file_hash,
-            'file_path': str(file_path),
-            'url': url,
-            'time': datetime.now().isoformat(),
-        })
-        return file_path
+        # If we have a cache hit and we do not want to redownload it and its still valid
+        # We want to just return the file
+        if cache_details and not redownload and not self._has_expired(cache_details):
+            return Path(cache_details['file_path'])
+        try:
+            file_path, file_hash, url = self._download_and_hash(urls, namespace)
+            # We explicitly check for redownload or expiry:
+            # This isn't needed as it should be caught bt the above if statement
+            # but this is more explicit
+            if cache_details and (redownload or self._has_expired(cache_details)):
+                self._storage.delete_by_key('url', cache_details['url'])
```

Comment:
I think there is logic problem here:

1. We do not remove any old files from disk
2. If the file changes on disk, we could remove it here but then it will break logic in the manager as we need to hash the file a second time.

Reply:
how about adding
 ``` 
os.remove(cache_details['file_path'])
```
removing the file and then deleting it?

Reply:
Will that work?

Reply:
It should since we would remove the file so no we would not have any old file remaining

Reply:
But what about the need to hash and update the database again?

Reply:
how about we check the file hash in cache and downloaded and only remove if its not same 
```
    try:
        file_path, file_hash, url = self._download_and_hash(urls, namespace)

        if cache_details and not redownload:

            if cache_details['file_hash'] == file_hash:
                return Path(cache_details['file_path'])

            os.remove(cache_details['file_path'])
            self._storage.delete_by_key('url', cache_details['url'])
```

Reply:
Maybe but we first need a unit test to replicate this and work out the best solution from there.